xcm/xcm-simulator/src/lib.rs

```diff
@@ -187,30 +207,67 @@ macro_rules! decl_test_network {
 			}
 		}
 
+		/// Process any XCMs in the message buses.
+		///
+		/// Called automatically by `TestExt::execute_with`.
+		fn process_messages() -> $crate::XcmResult {
+			process_relay_messages()?;
+			process_para_messages()
```

Comment:
In the scenario where `ParaA` sends relay chain tokens to `ParaB`, there would be a para message firstly (ump), and then a relay message (dmp). This impl won't process the relay message in this case.

IMO the messages process should loop process until there is no more messages in *any* buses, like:

```rust
while has_messages_in_any_bus {
    process_relay_messages();
    process_para_messages();
}
```